\maketitle 
\listoftodos
\tableofcontents % This creates the table of contents; autofills Chapters, sub-chapters (aka "sections"), and page numbers.

% Add an asterisk before the {} to make an un-numbered chapter that won't be included in the ToC.
\chapter{Foreword}

\section{Goal}
This tutorial aims to improve your general understanding of bioinformatics through several methods:
\begin{itemize}
    \item Define technical terms commonly used in bioinformatics methods and found in the literature.
    \item Provide a collection of various useful resources, including...
    \begin{enumerate}
        \item Resources for finding tools, data, and background information that can help answer your research questions.
        \item Resources that explain details about bioinformatics concepts and techniques in beginner-friendly language.
    \end{enumerate}
    \item Demonstrate how Python can be used to answer your research questions by combining existing bioinformatics tools and automating repetitive or time-consuming tasks.
\end{itemize}

\section{How to use this book}
Most reference materials are consolidated in the appendices at the end of the book. The main text of this book is written in the right-hand margin. The left-hand margin contains special markers and important notes to the reader. The book can be used as a self-paced tutorial with the help of the markers described below.

------------------------------- 

\textit{Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas eu felis sodales, interdum purus nec, interdum ex. Integer at nunc \marginlabel{This is a margin label. I will write things here to further explain the main text, define jargon, etc.} ultricies, tempus nibh eget, egestas risus. Suspendisse aliquam, lorem at dictum accumsan, dolor elit euismod velit, a gravida risus libero non felis. Donec a tortor tempor, scelerisque sapien posuere, volutpat erat. Morbi in imperdiet velit. Vivamus sagittis, massa sit amet venenatis euismod, elit eros aliquam diam, molestie faucibus lectus nisi euismod erat. Nulla id dictum mi.}

\attention \textbf{The arrow that points to this line is an "Attention" marker. It will indicate key pieces of information that you should pay special attention to.} \textit{Curabitur egestas aliquam nisl, pharetra finibus mauris placerat nec. Nam in diam risus. Nulla mollis purus quis feugiat tristique. Vestibulum et sollicitudin ante, at sagittis ipsum. Nunc hendrerit ante sed massa semper eleifend.}

\seealso{This kind of annotation will reference appendix entries that you can consult for more-detailed information about the main text.} Ut ultrices eros velit, at faucibus ante rutrum eget. Pellentesque a molestie diam. Curabitur mattis dui a risus lacinia fringilla. Phasellus porttitor elit nec neque euismod, id ultrices elit lobortis. Aliquam molestie sem. Curabitur sit amet urna faucibus, vulputate arcu sit amet, fermentum ipsum. Nulla facilisi. Nam ullamcorper eget leo id malesuada.

\section{Pre-requisites}
\subsection*{Things to know}
\subsection*{Computer requirements}
\begin{itemize}
    \item At least 4GB RAM bare minimum, 16GB RAM if possible (basically, the more RAM, the better)
    \item Must be able to run macOS 10.12 Sierra, preferably macOS 10.13 High Sierra
    \item At least 500GB storage (ideally several TB)
\end{itemize}

\chapter{Source Code}
(Source code can be found at github.com/EthanGniot/LU-microbiome)

\chapter{Sample Information}
\todo{Still need to add information about the inflammatory bowel disease dataset}

While we are building the LU-microbiome pipeline, we will need a dataset that we can use to test our pipeline throughout the process and make sure it is working as intended. In order to do this, we must use a dataset that's already been analyzed so that we know what the results should look like. When we test our pipeline, if our results match the results of the original analysis, then we will know that our tool is working correctly.

There are several test datasets available to us for testing the feature that analyzes the relative abundance of bacteria in the microbiota.

The first is the dataset used in both the QIIME ``Illumina Overview Tutorial" (\ref{appendix:IlluminaOverTut}) \todo{Make URLs footnotes instead of appendix entries} and the QIIME 2 ``Moving Pictures" tutorial (\ref{appendix:MovingPicTut}) derived from the \seealso{Caporaso et al, 2011 \cite{Caporaso2011}} \textit{Moving Pictures of the Human Microbiome} study, where two human subjects collected daily samples from four body sites: the tongue, the palm of the left hand, the palm of the right hand, and the gut (via fecal samples obtained by swapping used toilet paper). These data were sequenced using the barcoded amplicon sequencing protocol described in \seealso{ THIS CITATION NEEDS TO BE FIXED \cite{Caporaso2011a}} \textit{Global patterns of 16S rRNA diversity at a depth of millions of sequences per sample}. A more recent version of this protocol that can be used with the Illumina HiSeq 2000 and MiSeq can be found here. \todo{Fix bibliography entries in this paragraph and in general. They're not correctly referencing even though bib file has entries.}

(Here is information about the untested inflammatory bowel disease dataset that we will analyze using the completed pipeline)

\chapter{Setting the scene}
("Here" is a hypothetical situation/research question that a student may have. This is the research question that will be answered by the pipeline we are creating.)

\chapter{Microbiome Analysis}
(This chapter will give some general background information about the topics listed below. More-detailed information will be provided in later chapters when we are actually creating the pipeline)
\section{The Gut Microbiome}
\section{Relative Abundance Analysis}
\section{Metagenomics}
\section{Python}

\chapter{How to Find Tools}
\section{Finding Data}
(Here is where we talk about various databases that users can use to find general information, data files, study results, public datasets, etc.)
\section{Finding Software}
(Here is where we talk about ways/places that people can look for software programs that can help answer their research question.)
