\maketitle 
\listoftodos
\tableofcontents % This creates the table of contents; autofills Chapters, sub-chapters (aka "sections"), and page numbers.

% Add an asterisk before the {} to make an un-numbered chapter that won't be included in the ToC.
\chapter{Foreword}

\section{Goal}
This tutorial aims to improve your general understanding of bioinformatics through several methods:
\begin{itemize}
    \item Define technical terms commonly used in bioinformatics methods and found in the literature.
    \item Provide a collection of various useful resources, including...
    \begin{enumerate}
        \item Resources for finding tools, data, and background information that can help answer your research questions.
        \item Resources that explain details about bioinformatics concepts and techniques in beginner-friendly language.
    \end{enumerate}
    \item Demonstrate how Python can be used to answer your research questions by combining existing bioinformatics tools and automating repetitive or time-consuming tasks.
\end{itemize}

\section{How to Use this Manual}
The main text of this book is written in the right-hand margin, while the left-hand margin contains special markers and important notes to the reader. All of the resources referenced in the following chapters can be found in the left-hand margin, the appendices, or both. The book can be used as a self-paced tutorial with the help of the markers described below.

------------------------------- 

\marginlabel{This is a margin label. I will write things here to further explain the main text, define jargon, etc.} \textit{Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas eu felis sodales, interdum purus nec, interdum ex. Integer at nunc ultricies, tempus nibh eget, egestas risus.}

\vspace{8mm}

\attention{This is an ``Attenion" marker. It will indicate key pieces of information that you should pay special attention to.}\textit{Curabitur egestas aliquam nisl, pharetra finibus mauris placerat nec.}

\seealso{This kind of annotation will reference appendix entries that you can consult for more-detailed information about the main text.} Ut ultrices eros velit, at faucibus ante rutrum eget. Pellentesque a molestie diam. Curabitur mattis dui a risus lacinia fringilla. Phasellus porttitor elit nec neque euismod, id ultrices elit lobortis. 

------------------------------- 

Chapters 2 - 4 cover supplementary information about this project. Chapter 5 and Chapter 6 are educational in nature. Chapter 5 gives background information about microbiome analyses and references helpful resources for learning why these analyses are useful. Chapter 6 talks about various databases for finding public data, along with places where people can look for existing software programs that could help them answer their research questions. Chapter 7 breaks down the steps required to complete the microbiome analyses in this tutorial. 

Finally, Chapters 8 - 16 are about creating and using the BioPype pipeline. These chapters are hyper-specific to the BioPype commands, rather than bioinformatics in general, though they still provide resources for reading more about the topics that are discussed. Each chapter covers a different step of the microbiome analysis process, and each chapter is divided into two sections: a "How to Build" section and a "How to Use" section. 

The \textbf{How to Build} section teaches you how to build your own pipeline to answer your research questions. This is accomplished by explaining \textit{how} and \textit{why} the BioPype commands were created the way they were. By seeing the thought process behind BioPype's development process, you can adapt the process to suit your own development needs.

The \textbf{How to Use} section explains how to use BioPype's functionality to answer questions about the gut microbiome. The section provides a workflow that explains which commands should be used in which order to complete the step that the chapter focuses on. 

\section{Pre-requisites}
\subsection*{Things to know}
\todo[inline]{Talk about the biology concepts they should be familiar with before starting.}
\subsection*{Computer requirements}
\begin{itemize}
    \item At least 4GB RAM bare minimum, 16GB RAM if possible (basically, the more RAM, the better)
    \item Must be able to run macOS 10.12 Sierra, preferably macOS 10.13 High Sierra
    \item At least 500GB storage (ideally several TB)
\end{itemize}

\chapter{Source Code}
(Source code can be found at github.com/EthanGniot/BioPype)

\chapter{Sample Information}
\todo{Still need to add information about the inflammatory bowel disease dataset}

While we are building the BioPype pipeline, we will need a dataset that we can use to test our pipeline throughout the process and make sure it is working as intended. In order to do this, we must use a dataset that's already been analyzed so that we know what the results should look like. When we test our pipeline, if our results match those of the original analysis, then we will know that our tool is working correctly.

\section{Test Dataset: Relative Abundance Analysis}
There are several public datasets that can be used to test our code while we develop the microbial relative abundance analysis.

The first is the dataset used in both the QIIME ``Illumina Overview Tutorial" (\ref{appendix:IlluminaOverTut}) \todo{Make URLs footnotes instead of appendix entries} and the QIIME 2 ``Moving Pictures" tutorial (\ref{appendix:MovingPicTut}) derived from the \seealso{Caporaso et al, 2011 \cite{Caporaso2011}} \textit{Moving Pictures of the Human Microbiome} study, where two human subjects collected daily samples from four body sites: the tongue, the palm of the left hand, the palm of the right hand, and the gut (via fecal samples obtained by swapping used toilet paper). These data were sequenced using the barcoded amplicon sequencing protocol described in \seealso{ THIS CITATION NEEDS TO BE FIXED \cite{Caporaso2011a}} \textit{Global patterns of 16S rRNA diversity at a depth of millions of sequences per sample}.  \todo{Fix bibliography entries in this paragraph and in general. They're not correctly referencing even though bib file has entries.}

\todo[inline]{(Include information about the untested inflammatory bowel disease dataset that we will analyze using the completed pipeline)}

\chapter{Setting the scene}
("Here" is a hypothetical situation/research question that a student may have. This is the research question that will be answered by the pipeline we are creating.)

\chapter{Microbiome Analysis}
(This chapter will give some general background information about the topics listed below. More-detailed information will be provided in later chapters when we are actually creating the pipeline)
\section{The Gut Microbiome}
\section{Relative Abundance Analysis}
\section{Metagenomics}
\section{Python}

\chapter{How to Find Tools}
\section{Finding Data}
(Here is where we talk about various databases that users can use to find general information, data files, study results, public datasets, etc.)
\section{Finding Software}
(Here is where we talk about ways/places that people can look for software programs that can help answer their research question.)
