\maketitle 
\listoftodos
\tableofcontents % This creates the table of contents; autofills Chapters, sub-chapters (aka "sections"), and page numbers.

% Add an asterisk before the {} to make an un-numbered chapter that won't be included in the ToC.
\chapter{Foreword}

\section{Goal}
This tutorial aims to improve your general understanding of bioinformatics through several methods:
\begin{itemize}
    \item Define technical terms commonly used in bioinformatics methods and found in the literature.
    \item Provide a collection of various useful resources, including...
    \begin{enumerate}
        \item Resources for finding tools, data, and background information that can help answer your research questions.
        \item Resources that explain details about bioinformatics concepts and techniques in beginner-friendly language.
    \end{enumerate}
    \item Demonstrate how Python can be used to answer your research questions by combining existing bioinformatics tools and automating repetitive or time-consuming tasks.
\end{itemize}

\section{How to use this book}
The main text of this book is written in the right-hand margin. The left-hand margin contains special markers and important notes to the reader. Most reference materials are consolidated in the appendices at the end of the book. The book can be used as a self-paced tutorial with the help of the markers described below.

------------------------------- 

\textit{Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas eu felis sodales, interdum purus nec, interdum ex. Integer at nunc \marginlabel{This is a margin label. I will write things here to further explain the main text, define jargon, etc.} ultricies, tempus nibh eget, egestas risus. Suspendisse aliquam, lorem at dictum accumsan, dolor elit euismod velit, a gravida risus libero non felis. Donec a tortor tempor, scelerisque sapien posuere, volutpat erat. Morbi in imperdiet velit. Vivamus sagittis, massa sit amet venenatis euismod, elit eros aliquam diam, molestie faucibus lectus nisi euismod erat. Nulla id dictum mi.}

\attention \textbf{The arrow that points to this line is an "Attention" marker. It will indicate key pieces of information that you should pay special attention to.} \textit{Curabitur egestas aliquam nisl, pharetra finibus mauris placerat nec. Nam in diam risus. Nulla mollis purus quis feugiat tristique. Vestibulum et sollicitudin ante, at sagittis ipsum. Nunc hendrerit ante sed massa semper eleifend.}

\seealso{This kind of annotation will reference appendix entries that you can consult for more-detailed information about the main text.} Ut ultrices eros velit, at faucibus ante rutrum eget. Pellentesque a molestie diam. Curabitur mattis dui a risus lacinia fringilla. Phasellus porttitor elit nec neque euismod, id ultrices elit lobortis. Aliquam molestie sem. Curabitur sit amet urna faucibus, vulputate arcu sit amet, fermentum ipsum. Nulla facilisi. Nam ullamcorper eget leo id malesuada.


\section{Source Code}
The source code for this manual, the accompanying tutorial, and the BioPype package can all be found at \url{github.com/EthanGniot/BioPype}.
\section{Pre-requisites}
\subsection*{Computer requirements}
\begin{itemize}
    \item At least 4GB RAM minimum.* 16GB RAM if possible (the more RAM, the better)
    \begin{itemize}
    \item *At 4GB RAM, one step of the BioPype analysis takes \~2.5 hours to finish when analyzing 20 samples. With less RAM, expect the process to take longer. Conversely, more RAM will increase the speed of the analysis. 
    \end{itemize}
    \item BioPype was developed and tested using a MacBook Pro (Retina, 13-inch, Late 2013) running macOS 10.13.4 High Sierra. BioPype should be compatible with similar macOS versions, as well as Linux, though it has not yet been tested on Linux.
    \item Storage requirements will vary depending on the amount of data you intend to analyze. The analysis performed in this tutorial required at least 500GB and were satisfied by using a 2TB external hard drive.
\end{itemize}


%
%\chapter{Sample Information}

%While we are building the BioPype pipeline, we will need a dataset that we can use to test our pipeline throughout the process and make sure it is working as intended. In order to do this, we must use a dataset that's already been analyzed so that we know what the results should look like. When we test our pipeline, if our results match the results of the original analysis, then we will know that our tool is working correctly.

%There are several test datasets available to us for testing the feature that analyzes the relative abundance of bacteria in the microbiota.

%The first is the dataset used in both the QIIME ``Illumina Overview Tutorial" (\ref{appendix:IlluminaOverTut}) and the QIIME 2 ``Moving Pictures" tutorial (\ref{appendix:MovingPicTut}) derived from the \textit{Moving Pictures of the Human Microbiome} study, where two human subjects collected daily samples from four body sites: the tongue, the palm of the left hand, the palm of the right hand, and the gut (via fecal samples obtained by swapping used toilet paper) \cite{Caporaso2011}. These data were sequenced using the barcoded amplicon sequencing protocol described in \textit{Global patterns of 16S rRNA diversity at a depth of millions of sequences per sample} \cite{Caporaso2011a}. A more recent version of this protocol that can be used with the Illumina HiSeq 2000 and MiSeq can be found here.

%\chapter{Setting the scene}
%    \label{chap:scene}
%("Here" is a hypothetical situation/research question that a student may have. This is the research question that will be answered by the pipeline we are creating.)

\chapter{Microbiome Analysis}
The analysis used in this tutorial will focus on analysis of the human gut microbiome, but BioPype can be used to analyze the microbiomes of other communities as well.
\section{The Gut Microbiome}
%
\seealso{\footnotesize For more explanation about the \textbf{gut microbiota} and \textbf{gut microbiome}, see \ref{appendix:microbiome-vs-microbiota}.}
%
The human gut is home to trillions of microbes. The collection of physical micro-organisms that live in the gut is referred to as the \textbf{gut microbiota}. The \textbf{gut microbiome} refers to all of the genes contained within these micro-organisms, as well as their functions and interactions.

%
Human microbiota can play host to many different kinds of microbes, including bacteria, archaea, and fungi. For the sake of this tutorial, we are going to focus on the role of bacteria in human microbiomes.

\marginlabel{\footnotesize \textbf{Gut dysbiosis:} deviation from "normal" gut microbiota composition.}
Sometimes the biodiversity of the gut is thrown out of balance. When the composition of the microbiota is significantly altered, and bacterial populations that normally compose a large percentage of the microbiota become underrepresented (or underrepresented populations become overrepresented), the gut is in a state of \textit{dysbiosis}.
%
\seealso{\footnotesize For an approachable overview of the human microbiome, the effects of microbiome variation on human health and disease, and important research questions regarding the microbiome, see \cite{Cho2012}.}
%
 \textbf{Gut dysbiosis} is important to study because it has been implicated in many physical and neurological human health conditions, including diabetes, rheumatoid arthritis, Alzheimer's, inflammatory bowel disease (IBD), colon cancer \citep{Buford2017, Kennedy2014, Castellarin2012},  chronic fatigue syndrome \citep{Lakhan2010}, obesity \citep{Turnbaugh2006,Turnbaugh2009}, bacterial vaginosis \citep{Africa2014}, ulcerative colitis {\citep{Mandal2015, Matsuoka2015}, anxiety, and depression \citep{Evrensel2015, Lach2018}
%
\section{Relative Abundance Analysis}
%
\marginlabel{\footnotesize \textbf{Relative abundance} is a measure of how common or rare a taxon is, relative to other taxa in its community, and is usually expressed as a percentage. For more information on \textbf{relative abundance} analyses, see Appendix \ref{appendix:abundance-analyses}.} 
%
BioPype lets users examine the \textbf{relative abundance} of bacterial taxa in each experimental sample. Relative abundance analyses help visualize the structure of the micro-organism community. Users can use BioPype to analyze bacterial \textbf{amplicon} sequencing data to determine which micro-organisms are present and in what proportion.
%
\seealso{\footnotesize For information on \textbf{amplicon} sequencing vs. shotgun sequencing, see Appendix \ref{appendix:amplicon}.}
%
The idea is to take the sequences from a gut sample and assign them to a taxon. To do that, we group (or cluster) sequences based on their similarity to define \textbf{Operational Taxonomic Units (OTUs)}. 
%
\marginlabel{\footnotesize \textbf{Operational Taxonomic Unit (OTU)}: A group of sequences that share significant similarity based on some marker gene. Groups can be treated as a single taxon (e.g., genus, species, etc. depending on the clustering threshold) for analysis purposes. For more info, see Appendix \ref{appendix:otu}.}
%
Using the number of sequences assigned to each OTU/taxon, BioPype can determine the relative abundance of each micro-organism in a sample and plot the results on an interactive bar chart.

For example, studies have found that the genera \textit{Bacteroides} and \textit{Lactobacillus} are both prevalent in the healthy human gut microbiome, but \textit{Bacteroides} has a higher relative abundance than \textit{Lactobacillus} \cite{Lloyd-Price2016}. In other words, out of all the sequences obtained from a healthy human gut sample, a larger percentage of those sequences will be assigned to the \textit{Bacteroides} OTU than to the \textit{Lactobacillus} OTU during a relative abundance analysis. 

Relative abundance is an important measure of biodiversity in the gut. Using BioPype to perform a relative abundance analysis allows examination of gut biodiversity at 7 different taxonomic levels, which helps identify instances of gut dysbiosis in research subjects. 

\section{Metagenomics}
\attention{\textit{THIS FUNCTIONALITY HAS NOT BEEN COMPLETED YET, BUT WILL BE ADDED IN FUTURE UPDATES}}

BioPype can also be used to predict which genes are present in microbiome samples. From these predicted genes, users can do two things. 

First, they can generalize the functions encoded across all bacteria that are present in an experimental condition's microbiome (i.e., at the metagenomic scale) and visualize the relative distribution of these functions with a bar chart. Having this functional snapshot of a metagenome is useful because a functional overview of the experimental condition can be compared to a functional overview of the control condition to see if there are differences in the two groups' functional capabilities. When comparing the microbiomes of disordered vs. non-disordered samples, having a visual depiction of the differences in functional capabilities can make it easier to find the cause of the disorder by narrowing down the potential sources of the condition. 

For example, certain G protein-coupled receptors (GPCRs) in the gut are associated with some of the diseases associated with gut dysbiosis, such as inflammatory bowel disease (IBD) \citep{Cohen2017}. Research has also shown that bacteria interact with their environment via the release of small molecules \citep{Li2012}. From these premises, one could hypothesize that IBD may be caused by the over-production of certain effector molecules from the gut microbiome. This hypothesis may be tested by using BioPype to compare the functional capabilities of the IBD microbiome with that of a healthy control microbiome. If the results show, for example, that the IBD gut metagenome has a greater proportion of genes associated with lipid production than the control metagenome, it indicates potential support for the user's hypothesis. They can now refine their investigation by narrowing the scope of their search to only analyze genes associated with lipid synthesis.

Second, users can use BioPype to predict the functions of \textit{individual} genes. This is useful because it allows users to identify a specific gene of interest for their research question, which they can then design a validation experiment for. 

Continuing with our above example hypothesis about the gut microbiome's role in IBD, the user can now turn their focus to the functions of specific lipid-synthesis genes that are being overrepresented in IBD patients' microbiomes. Upon examining the functions of individual genes, they find that N-acyl amide synthase genes are enriched in the IBD microbiome. Wet-lab validation experiments will confirm that the lipids encoded by these genes are able to interact with GPCRs that regulate GI tract physiology \citep{Cohen2017}. This suggests that these lipids, and their associated genes in the microbiome, have the potential to cause IBD during times of dysbiosis by activating GPCRs in the gut \citep{Cohen2017}, thus supporting the user's original hypothesis. Additionally, one can use these findings to focus future research questions about the role of the human microbiome in health disorders by searching for correlations between N-acyl amide synthase genes and other conditions. For instance, since the products of these genes interact with GPCRs to potentially regulate IBD, and IBD is associated with anxiety and depression, investigation may reveal that the interaction between these genes and GPCRs potentially regulate anxiety and depression as well.

%
\section{Python}
To make use of this tutorial and the BioPype package, you will need to be familiar with the basics of the Python coding language. The free DataCamp course \textit{Intro to Python for Data Science} (\url{https://www.datacamp.com/courses/intro-to-python-for-data-science?tap_a=5644-dce66f&tap_s=116411-750171}) is a quick and easy introduction to the basics of the language. It covers:
%
\begin{itemize}
\item Variables and variable assignment
\item Basic data types (strings, integers, floats, lists, booleans, ) and converting between types
\item Calculations with variables
\item Functions and arguments
\item The \verb|help()| function.
\item Methods
\item Packages and importing packages
\item The basics of the NumPy package (The material covered in the NumPy chapter of the DataCamp tutorial is not required for using BioPype, but if you are new to Python,  it's a great way to practice/apply the concepts they teach you in Chapters 1-3.)
\end{itemize}
%
The DataCamp course should take \~4 hours to complete (less, if you forego NumPy chapter), and will teach you everything you'll need to know about Python in order to make use of BioPype. 

If you decide that you'd like to start creating your own tools with Python, there's a bit more you'll need to know. Resources for learning how to create your own functions, classes, packages, and more can be found in Appendix \ref{chap:python-resources}.


\section{BioPype Tutorial Dataset}

The dataset we will analyze in this tutorial comes from the \textit{Longitudinal Multi'omics of the Human Microbiome in Inflammatory Bowel Disease} study (unpublished) (see \url{https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP115494}).

From the abstract: "The main Inflammatory Bowel Disease (IBD) Multi''omics Database (IBDMDB) study includes multi'omics measurements from over 100 subjects, sampled biweekly over up to a year in both adult and pediatric patients with IBD (Crohn's disease and ulcerative colitis), along with non-IBD controls. Data types include fecal metagenomes, metatranscriptomes, metabolomes, and proteomes, as well as host genetics, intestinal biopsy transcriptomes, epigenetics, and 16S amplicon profiles. Subjects' medical histories and demographics are collected at baseline and medication, diet, and disease activity profiled longitudinally." 

In our analysis, we will be downloading sequencing data from their NCBI Sequence Read Archive page that corresponds to patients with ulcerative colitis and Crohn's disease (both conditions fall under the umbrella condition of "inflammatory bowel disease". We will examine whether there are compositional differences between the gut microbiota of younger patients with ulcerative colitis or Crohn's disease compared with their older cohort. While we expect both groups to exhibit gut dysbiosis and clear deviation from the composition of a healthy human gut microbiome, what we're hoping to study is whether or not increased age exacerbates the severity of the dysbiosis. Increased age is associated with gut dysbiosis, even in otherwise healthy individuals, so we predict that older IBD patients will exhibit greater degrees of dysbiosis than their younger counterparts. 


%\chapter{How to Find Tools}
%    \label{chap:find-tools}
%\section{Finding Data}
%(Here is where we talk about various databases that users can use to find general information, data files, study results, public datasets, etc.)
%\section{Finding Software}
%(Here is where we talk about ways/places that people can look for software programs that can help answer their research question.)
